RFD.md

```diff
@@ -0,0 +1,217 @@
+
+#### Availability and resource limits
+
+In a multi-tenant environment like the system in question, we need to ensure proper resource limits to reduce the potential impact of noisy or abusive clients on the underlying host OS.
+
+We're going to use linux Cgroups to apply the following limits on each command:
+
+* Memory usage – a static per-command limit will be applied on each container.
+
+* CPU usage - each command will be limited to a small constant amount of CPU resources, preventing it from negative affecting the host OS and other commands by consuming too much CPU.
+
+* Disk IO - each command will be limited by the amount of disk IO it could perform to limit the effects from a single command's high IO load on the host OS and other commands.
+
+##### Scope limits
+
+All limits will be statically hardcoded within the server, while in a production system we'd probably have some kind of configuration file for default limits and would expand the API to allow users to specify their own limits for each command (like Docker does).
+
+In a production environment, we may want to introduce per-user resource limits, API call rate limits, etc to further protect the system from abuse.
+
+### API
+
+The following API methods will be available on the server:
+
+* `Status` - returns some basic status information
+* `StartCommand` - starts a command on the server and returns a unique `command_id` value used to manage the command in subsequent API calls.
+* `StopCommand` - stops a given command (identified by a `command_id`).
+* `CommandStatus` - returns current status information for a given command (identified by a `command_id`).
+* `CommandOutput` - a streaming API for receiving console output from a given command (identified by a `command_id`). If requested, the stream will continue until the command has finished (tail mode).
+
+The detailed GRPC definition for the proposed API can be found within the `remote_exec.proto` file.
+
+### Library design
+
+Most of concerns around containerization will be isolated within the `container_exec` library, allowing us to have a very simple GRPC server around the library. This should make it easier to test the containerization code and contain most of the complexity (dealing with Linux APIs, process management, logs management, concurrency concerns, etc) the library.
+
+In a production scenario, this approach would make it possible to potentially reuse the library in multiple different systems (see ContainerD core services or the Moby project and its usage within Docker).
+
+#### Scope limits and design trade-offs
+
+We're going to apply the following design decisions while building the library:
+
+* Just like in unix process management, we keep finished command status and command logs only until someone calls `Status()` on it (think of it just like a zombie process).
```

Comment:
I wouldn't do this. It could allow a user to "hide" their command from an admin by calling `Status` as soon as possible to purge any record of the job running.

Reply:
Alternatively, we can leave the information around forever and make deletion explicit like it is done in Docker API (where you need to specify `--rm` in `docker run` or have to remove the container manually later). Now that I think about it, it looks cleaner indeed.

Reply:
Agree that keeping it forever opens the server up to DoS attacks, but as long as you mention that in the doc I think it's fine. 

Explicit delete like docker is a fine solution but that seems like busy work and is not required as part of the challenge.